File: solutions/300.longest-increasing-subsequence/longest-increasing-subsequence.py
class Solution(object):
    def lengthOfLIS(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        if not nums:
            return 0
        size = 0
        tails = []
        for n in nums:
            i = bisect.bisect_left(tails, n)
            if i >= len(tails):
                tails.append(n)
            else: 
                tails[i] = n
            size = max(i + 1, size)
        return size

File: solutions/300.longest-increasing-subsequence/longest-increasing-subsequence.java
class Solution {
    public int lengthOfLIS(int[] nums) {
        if (nums.length == 0) {
            return 0;
        }
        int len = nums.length;
        int[] tails = new int[len];

        int size = 0;
        for (int n : nums) {
            int s = 0, e = size;
            while (s < e) {
                int m = s + (e-s) / 2;
                if (n > tails[m]) {
                    s = m + 1;
                } else if (n < tails[m])
                    e = m;
                else {
                    e = m;
                }
            }
            tails[s] = n;
            size = Math.max(s + 1, size);
        }
        return size;
    }
}